kpops/utils/yaml_loading.py

```diff
@@ -37,3 +39,54 @@ def substitute(input: str, substitution: Mapping[str, Any] | None = None) -> str
     if not substitution:
         return input
     return Template(input).safe_substitute(**substitution)
+
+
+def substitute_nested(input: str, max_repetitions: int = 500, **kwargs) -> str:
+    """Allow for multiple substitutions to be passed.
+
+    Will make as many passes as needed to substitute all possible placeholders.
+    A ceiling is set to avoid infinite loops.
+
+    HINT: If :param input: is a ``Mapping`` that you converted into ``str``,
+    You can pass it as a string, and as a ``Mapping`` to enable self-reference.
+
+    :Example:
+
+    >>> substitution = {
+            "a": "0",
+            "b": "${a}",
+            "c": "${b}",
+            "d": "${a}",
+        }
+    >>> input = "${a}, ${b}, ${c}, ${d}"
+    >>> print("Substituted string: " + substitute_nested(input, **substitution))
+    0, 0, 0, 0
+
+    :param input: The raw input containing $-placeholders
+    :type input: str
+    :param max_repetitions: The max number of consecutive substitutions to be carried out, defaults to 500
+    :type max_repetitions: int, optional
+    :param **kwargs: Substitutions
+    :raises Exception: Substitution was repeated {counter} times and placeholders still exist, check for loops
+    :return: Substituted input string
+    :rtype: str
+    """
+    if not kwargs:
+        return input
+    old_str, new_str = "", substitute(input, kwargs)
+    counter = 0
+    starting_time = time.time()
+    while old_str != new_str:
+        old_str, new_str = new_str, substitute(new_str, kwargs)
+        counter += 1
+        current_time = time.time()
+        if current_time - starting_time > 10:
+            logging.warn(
+                f"Substitution has been repeating for 10 seconds, possible loop detected. The program will exit after reaching {max_repetitions} tries."
```

Comment:
let's have a think once more what are the exact conditions for a doom loop occuring and see if we can come up with a better detection

Reply:
[related test](https://github.com/bakdata/kpops/pull/198/files#diff-c06c7ddac9f17bd76f0e58bfc3a019ae647f7a305534a96b15e80c83520f777cR167-R185)

[Recognized conditions](https://github.com/bakdata/kpops/pull/198/files#diff-b67554f5da44782699825ea328c85f262297f18162ac18cc10504e8f2e7dc21d)